tools/rewriter/GenCallAsm.cpp

```diff
@@ -11,125 +12,168 @@
-        locs.push_back(ParamLocation::Stack());
+        arg.allocate_stack(stack_offset);
+        args.push_back(arg);
+        stack_offset += arg.size();
       }
       break;
     }
-    case CAbiArgKind::Memory: {
-      locs.push_back(ParamLocation::Stack());
+    case ArgLocation::Kind::Memory: {
+      assert(arg.is_stack());
+      if (stack_offset % arg.align() != 0) {
+        stack_offset += arg.align() - stack_offset % arg.align();
+      }
+      arg.allocate_stack(stack_offset);
+      args.push_back(arg);
+      stack_offset += arg.size();
+      if (stack_offset % 8 != 0) {
+        stack_offset += 8 - stack_offset % 8;
+      }
       break;
     }
     }
   }
-  return locs;
+  *stack_arg_size_out = stack_offset;
+  return args;
 }
 
-std::vector<ParamLocation> return_locations(const CAbiSignature &func) {
-  std::vector<ParamLocation> locs = {};
-
-  if (func.ret.empty()) {
-    return locs;
-  }
+std::vector<ArgLocation> allocate_return_locations(const AbiSignature &func, Arch arch) {
+  const auto &int_ret_reg_order = (arch == Arch::X86) ? x86_int_ret_reg_order : arm_int_ret_reg_order;
+  const auto &float_reg_order = (arch == Arch::X86) ? x86_float_reg_order : arm_float_reg_order;
 
+  std::vector<ArgLocation> args;
   size_t ints_used = 0;
   size_t floats_used = 0;
-  for (const auto &kind : func.ret) {
-    switch (kind) {
-    case CAbiArgKind::Integral:
-      assert(ints_used < 2);
-      locs.push_back(ParamLocation::Register(int_ret_reg_order[ints_used]));
-      ints_used += 1;
-      break;
-    case CAbiArgKind::Float:
+  for (auto arg : func.ret) {
+    switch (arg.kind()) {
+    case ArgLocation::Kind::Integral:
+      if (arg.size() <= 8 || arg.is_indirect()) {
+        assert(ints_used < int_ret_reg_order.size());
+        arg.allocate_reg(int_ret_reg_order[ints_used]);
+        args.push_back(arg);
+        ints_used += 1;
+        break;
+      } else {
+        assert(arg.size() == 16);
+        assert(!arg.is_indirect());
+        assert(ints_used + 1 < int_ret_reg_order.size());
+        for (int i = 0; i < 2; i++) {
+          auto half_arg = ArgLocation::Register(ArgLocation::Kind::Integral, 8);
+          half_arg.allocate_reg(int_ret_reg_order[ints_used]);
+          args.push_back(half_arg);
+          ints_used += 1;
+        }
+      }
+    case ArgLocation::Kind::Float:
       // TODO: handle x87 in st0 and complex x87 in st0+st1
-      assert(floats_used < 2);
-      locs.push_back(ParamLocation::Register(xmms[floats_used]));
+      arg.allocate_reg(float_reg_order[floats_used]);
+      args.push_back(arg);
       floats_used += 1;
       break;
-    case CAbiArgKind::Memory:
-      // memory return also returns address in first return register
-      if (locs.empty()) {
-        locs.push_back(ParamLocation::Register(int_ret_reg_order[0]));
+    case ArgLocation::Kind::Memory:
+      // memory return also returns address in first return register on x86
+      if (arch == Arch::X86) {
```

Comment:
Maybe assert `arch == Arch::X86` here?

Reply:
We can have memory returns on ARM, so I don't think this can be an assert. We just don't use one of the regular return registers to point to the address, there's a dedicated register.

Reply:
If there's a memory return (`func.ret` has one item w/ `ArgLocation::Kind::Memory`) then `allocate_return_locations` returns an empty vector and `stack_return_size` in `emit_asm_wrapper` stays 0 right? This seems like an issue unless I'm missing something.

Reply:
Good point. All memory returns were classified as indirect on AArch64, but that resulted in us preserving an extra return register as well, which wasn't necessary. I've clarified and fixed this up now.